File: packages/sdk/src/collectors/openai.ts
import type { LanternTracer } from "../tracer.js";
import type { SpanOutput } from "../types.js";

/**
 * Shape of the OpenAI client we wrap.
 * We use structural types so we don't need the actual OpenAI SDK as a dependency.
 */
interface OpenAIMessage {
  role: string;
  content: string | null;
}

interface OpenAIChoice {
  message: {
    role: string;
    content: string | null;
    tool_calls?: Array<{
      id: string;
      type: string;
      function: { name: string; arguments: string };
    }>;
  };
  finish_reason: string;
}

interface OpenAIUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

interface OpenAIResponse {
  id: string;
  choices: OpenAIChoice[];
  model: string;
  usage: OpenAIUsage;
}

interface OpenAICreateParams {
  model: string;
  messages: OpenAIMessage[];
  max_tokens?: number;
  temperature?: number;
  tools?: unknown[];
  [key: string]: unknown;
}

interface OpenAICompletions {
  create(params: OpenAICreateParams): Promise<OpenAIResponse>;
}

interface OpenAIClient {
  chat: {
    completions: OpenAICompletions;
  };
}

/**
 * Wrap an OpenAI client to automatically trace all chat.completions.create() calls.
 * Creates llm_call spans with full input/output/token data.
 *
 * Usage:
 * ```typescript
 * import OpenAI from "openai";
 * import { LanternTracer, wrapOpenAIClient } from "@lantern-ai/sdk";
 *
 * const client = wrapOpenAIClient(new OpenAI(), tracer);
 * // All client.chat.completions.create() calls are now traced
 * ```
 */
export function wrapOpenAIClient<T extends OpenAIClient>(
  client: T,
  tracer: LanternTracer,
  opts?: { traceId?: string; agentName?: string }
): T {
  const originalCreate = client.chat.completions.create.bind(client.chat.completions);

  const wrappedCreate = async (params: OpenAICreateParams): Promise<OpenAIResponse> => {
    // Get or create a trace
    let traceId = opts?.traceId;
    let ownTrace = false;
    if (!traceId) {
      const trace = tracer.startTrace({
        agentName: opts?.agentName ?? "openai-agent",
      });
      traceId = trace.id;
      ownTrace = true;
    }

    // Create the LLM call span
    const messages = params.messages.map((m) => ({
      role: m.role,
      content: m.content ?? "",
    }));

    const span = tracer.startSpan(traceId, {
      type: "llm_call",
      input: { messages },
      model: params.model,
    });

    try {
      const response = await originalCreate(params);

      // Extract text content from the first choice (primary response)
      const firstChoice = response.choices[0];
      const textContent = firstChoice?.message?.content ?? "";

      // Extract tool calls if present
      const toolCalls = firstChoice?.message?.tool_calls?.map((tc) => ({
        id: tc.id,
        name: tc.function.name,
        input: tc.function.arguments,
      })) ?? [];

      const output: SpanOutput = {
        content: textContent,
        toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
        stopReason: firstChoice?.finish_reason ?? undefined,
      };

      tracer.endSpan(span.id, output, {
        inputTokens: response.usage.prompt_tokens,
        outputTokens: response.usage.completion_tokens,
      });

      // Create child spans for tool calls
      for (const tool of toolCalls) {
        const toolSpan = tracer.startSpan(traceId, {
          type: "tool_call",
          parentSpanId: span.id,
          input: { args: tool.input },
          toolName: tool.name,
        });

        // Tool call spans are started but not ended here —
        // the actual tool execution happens outside this wrapper.
        // The user can end them manually or they'll be captured
        // by MCP instrumentation.
        tracer.endSpan(toolSpan.id, { content: "Tool call initiated" });
      }

      if (ownTrace) {
        tracer.endTrace(traceId, "success");
      }

      return response;
    } catch (error) {
      tracer.endSpan(span.id, {}, {
        error: error instanceof Error ? error.message : String(error),
      });

      if (ownTrace) {
        tracer.endTrace(traceId, "error");
      }

      throw error;
    }
  };

  // Replace the create method with the wrapped version
  Object.assign(client.chat.completions, { create: wrappedCreate });

  return client;
}

File: sdk-python/src/lantern_ai/collectors/openai.py
"""Auto-instrumentation for the OpenAI Python SDK.

Usage::

    from openai import OpenAI
    from lantern_ai import LanternTracer, wrap_openai_client

    tracer = LanternTracer(...)
    client = OpenAI()
    wrap_openai_client(client, tracer)

    # All client.chat.completions.create() calls are now traced automatically.
"""

from __future__ import annotations

import functools
from typing import Any, Optional, TYPE_CHECKING

from ..types import SpanInput, SpanOutput

if TYPE_CHECKING:
    from ..tracer import LanternTracer


def wrap_openai_client(
    client: Any,
    tracer: "LanternTracer",
    *,
    trace_id: Optional[str] = None,
    agent_name: Optional[str] = None,
) -> Any:
    """Monkey-patch an OpenAI client so every ``chat.completions.create()``
    call is automatically recorded as an ``llm_call`` span.

    Handles both the sync (``OpenAI``) and async (``AsyncOpenAI``) clients.
    The original response object is returned unchanged.

    Parameters
    ----------
    client:
        An ``openai.OpenAI`` or ``openai.AsyncOpenAI`` instance.
    tracer:
        The :class:`LanternTracer` to record spans on.
    trace_id:
        If provided, spans are added to this existing trace. Otherwise a
        new trace is created per call.
    agent_name:
        Agent name to use when auto-creating traces (default
        ``"openai-agent"``).

    Returns
    -------
    The same *client* instance (mutated in place).
    """

    original_create = client.chat.completions.create

    import asyncio

    is_async = asyncio.iscoroutinefunction(original_create)

    def _extract_messages(params: dict) -> list:
        messages = []
        for m in params.get("messages", []):
            content = m.get("content") or ""
            messages.append({"role": m.get("role", ""), "content": str(content)})
        return messages

    def _build_output(response: Any) -> tuple:
        """Return (SpanOutput, tool_calls_list)."""
        choices = getattr(response, "choices", [])
        first = choices[0] if choices else None
        message = getattr(first, "message", None) if first else None

        text_content = getattr(message, "content", "") or "" if message else ""
        finish_reason = getattr(first, "finish_reason", None) if first else None

        tool_calls = []
        raw_tool_calls = getattr(message, "tool_calls", None) if message else None
        if raw_tool_calls:
            for tc in raw_tool_calls:
                fn = getattr(tc, "function", None)
                tool_calls.append({
                    "id": getattr(tc, "id", None),
                    "name": getattr(fn, "name", None) if fn else None,
                    "input": getattr(fn, "arguments", None) if fn else None,
                })

        output = SpanOutput(
            content=text_content,
            tool_calls=tool_calls if tool_calls else None,
            stop_reason=finish_reason,
        )
        return output, tool_calls

    if is_async:

        @functools.wraps(original_create)
        async def _wrapped_create_async(*args: Any, **kwargs: Any) -> Any:
            params = kwargs

            tid = trace_id
            own_trace = False
            if tid is None:
                t = tracer.start_trace(agent_name=agent_name or "openai-agent")
                tid = t.id
                own_trace = True

            messages = _extract_messages(params)
            span = tracer.start_span(
                tid,
                type="llm_call",
                input=SpanInput(messages=messages),
                model=params.get("model", "unknown"),
            )

            try:
                response = await original_create(*args, **kwargs)

                output, tool_calls = _build_output(response)
                usage = getattr(response, "usage", None)
                tracer.end_span(
                    span.id,
                    output,
                    input_tokens=getattr(usage, "prompt_tokens", None),
                    output_tokens=getattr(usage, "completion_tokens", None),
                )

                for tool in tool_calls:
                    tool_span = tracer.start_span(
                        tid,
                        type="tool_call",
                        parent_span_id=span.id,
                        input=SpanInput(args=tool.get("input")),
                        tool_name=tool.get("name"),
                    )
                    tracer.end_span(
                        tool_span.id,
                        SpanOutput(content="Tool call initiated"),
                    )

                if own_trace:
                    tracer.end_trace(tid, "success")

                return response

            except Exception as exc:
                tracer.end_span(
                    span.id,
                    SpanOutput(),
                    error=str(exc),
                )
                if own_trace:
                    tracer.end_trace(tid, "error")
                raise

        client.chat.completions.create = _wrapped_create_async

    else:

        @functools.wraps(original_create)
        def _wrapped_create_sync(*args: Any, **kwargs: Any) -> Any:
            params = kwargs

            tid = trace_id
            own_trace = False
            if tid is None:
                t = tracer.start_trace(agent_name=agent_name or "openai-agent")
                tid = t.id
                own_trace = True

            messages = _extract_messages(params)
            span = tracer.start_span(
                tid,
                type="llm_call",
                input=SpanInput(messages=messages),
                model=params.get("model", "unknown"),
            )

            try:
                response = original_create(*args, **kwargs)

                output, tool_calls = _build_output(response)
                usage = getattr(response, "usage", None)
                tracer.end_span(
                    span.id,
                    output,
                    input_tokens=getattr(usage, "prompt_tokens", None),
                    output_tokens=getattr(usage, "completion_tokens", None),
                )

                for tool in tool_calls:
                    tool_span = tracer.start_span(
                        tid,
                        type="tool_call",
                        parent_span_id=span.id,
                        input=SpanInput(args=tool.get("input")),
                        tool_name=tool.get("name"),
                    )
                    tracer.end_span(
                        tool_span.id,
                        SpanOutput(content="Tool call initiated"),
                    )

                if own_trace:
                    tracer.end_trace(tid, "success")

                return response

            except Exception as exc:
                tracer.end_span(
                    span.id,
                    SpanOutput(),
                    error=str(exc),
                )
                if own_trace:
                    tracer.end_trace(tid, "error")
                raise

        client.chat.completions.create = _wrapped_create_sync

    return client
